lib/sys-info.ts

```diff
@@ -0,0 +1,56 @@
+import { ChildProcess } from "./wrappers/child-process";
+import * as Promise from "bluebird";
+import * as path from "path";
+
+export class SysInfo {
+	// Different java has different format for `java -version` command.
+	private static JAVA_VERSION_REGEXP = /(?:openjdk|java) version \"((?:\d+\.)+(?:\d+))/i;
+
+	// For other versions of java javac version output is not on first line.
+	// Thus can't use ^ for starts with in regex.
+	private static JAVA_COMPILER_VERSION_REGEXP = /javac (.*)/i;
+
+	private javaVerCache: string;
+	private javaCompilerVerCache: string;
+
+	constructor(private childProcess: ChildProcess) { }
+
+	public getJavaVersion(): Promise<string> {
+		if (!this.javaVerCache) {
+			return new Promise<string>((resolve, reject) => {
```

Comment:
Why do you need this new Promise, can't you just return the result of spawnFromEvent:
```TypeScript
return this.childProcess.spawnFromEvent("java", ["-version"], "exit")
					.then((spawnResult) => {
						this.javaVerCache = SysInfo.JAVA_VERSION_REGEXP.exec(spawnResult.stderr)[1];
					})
					.catch((err) => {
						this.javaVerCache = null;
					}).then(() => {
						return this.javaVerCache;
					});
```